x/projects/keeper/project.go

```diff
@@ -48,32 +49,115 @@ func (k Keeper) GetProjectForDeveloper(ctx sdk.Context, developerKey string, blo
-		return utils.LavaFormatWarning("could not add keys to project", fmt.Errorf("project not found"),
+	if found := k.projectsFS.FindEntry(ctx, projectID, ctxBlock, &project); !found {
+		return utils.LavaFormatWarning("failed to add keys",
+			fmt.Errorf("project not found"),
 			utils.Attribute{Key: "project", Value: projectID},
 		)
 	}
 
-	// check if the admin key is valid
 	if !project.IsAdminKey(adminKey) {
-		return utils.LavaFormatWarning("could not add keys to project", fmt.Errorf("the requesting key is not admin key"),
+		return utils.LavaFormatWarning("failed to add keys",
+			fmt.Errorf("requesting key must be admin key"),
+			utils.Attribute{Key: "project", Value: projectID},
+		)
+	}
+
+	// make changes on the future state
+	project = types.Project{}
+	if found := k.projectsFS.FindEntry(ctx, projectID, nextEpoch, &project); !found {
+		return utils.LavaFormatError("failed to add keys",
+			fmt.Errorf("project not found (for next epoch)"),
 			utils.Attribute{Key: "project", Value: projectID},
 		)
 	}
 
 	for _, projectKey := range projectKeys {
-		err := k.registerKey(ctx, projectKey, &project, uint64(ctx.BlockHeight()))
+		err := k.registerKey(ctx, projectKey, &project, nextEpoch)
 		if err != nil {
 			return utils.LavaFormatError("failed to register key to project", err,
 				utils.Attribute{Key: "project", Value: projectID},
-				utils.Attribute{Key: "key", Value: projectKey.Key},
-				utils.Attribute{Key: "keyTypes", Value: projectKey.Kinds},
 			)
 		}
 	}
 
-	return k.projectsFS.AppendEntry(ctx, projectID, uint64(ctx.BlockHeight()), &project)
+	return k.projectsFS.AppendEntry(ctx, projectID, nextEpoch, &project)
+}
+
+func (k Keeper) DelKeysFromProject(ctx sdk.Context, projectID string, adminKey string, projectKeys []types.ProjectKey) error {
+	ctxBlock := uint64(ctx.BlockHeight())
+
+	nextEpoch, err := k.epochstorageKeeper.GetNextEpoch(ctx, ctxBlock)
+	if err != nil {
+		return utils.LavaFormatError("DelKeysFromProject: failed to get NextEpoch", err,
+			utils.Attribute{Key: "index", Value: projectID},
+		)
+	}
+
+	// validate the admin key against the current project state; but make the
+	// changes on the future (= end of epoch) version if already exists. This
+	// ensures that we do not lose changes if there are e.g. multiple additions
+	// in the same epoch.
+
+	// validate admin key with current state
+	var project types.Project
+	if found := k.projectsFS.FindEntry(ctx, projectID, ctxBlock, &project); !found {
+		return utils.LavaFormatWarning("failed to delete keys",
+			fmt.Errorf("project not found"),
+			utils.Attribute{Key: "project", Value: projectID},
+		)
+	}
+
+	if !project.IsAdminKey(adminKey) {
+		return utils.LavaFormatWarning("failed to delete keys",
+			fmt.Errorf("requesting key must be admin key"),
+			utils.Attribute{Key: "project", Value: projectID},
+		)
+	}
+
+	project = types.Project{}
+	if found := k.projectsFS.FindEntry(ctx, projectID, nextEpoch, &project); !found {
```

Comment:
if the project isnt found in the next epoch than create one.
the project only reason to have a fixation in the future is if a change has happened